irods_capability_automated_ingest/sync_irods.py

```
@@ -120,18 +124,129 @@ def register_file(hdlr_mod, logger, session, meta, **options):
 
+def get_s3_client(meta):
+
+    proxy_url = meta.get('s3_proxy_url')
+    if proxy_url is None:
+        httpClient = None
+    else:
+        import urllib3
+        httpClient = urllib3.ProxyManager(
+                                proxy_url,
+                                timeout=urllib3.Timeout.DEFAULT_TIMEOUT,
+                                cert_reqs='CERT_REQUIRED',
+                                retries=urllib3.Retry(
+                                    total=5,
+                                    backoff_factor=0.2,
+                                    status_forcelist=[500, 502, 503, 504]
+                                )
+                     )
+    endpoint_domain = meta.get('s3_endpoint_domain')
+    s3_region_name = meta.get('s3_region_name')
+    s3_access_key = meta.get('s3_access_key')
+    s3_secret_key = meta.get('s3_secret_key')
+    s3_secure_connection = meta.get('s3_secure_connection')
+    if s3_secure_connection is None:
+        s3_secure_connection = True
+
+    isLocalHost = ['localhost', '127.0.0.1']
+    #Minio server running locally causes SSL error if secure set to True
+    if endpoint_domain.split(":")[0] in isLocalHost:
+        s3_secure_connection = False
+
+    client = minio.Minio(
+                 endpoint_domain,
+                 region=s3_region_name,
+                 access_key=s3_access_key,
+                 secret_key=s3_secret_key,
+                 secure=s3_secure_connection,
+                 http_client=httpClient)
+    return client
+
+def upload_file_from_S3(logger, session, meta, src, dest, offset=0, **options):
+
+    #Could pass in as parameter when running ingest job, keep in meta
+    #May be needed for files uploaded to AWS S3 with multi-part upload
+    BUFFER_SIZE = 1024
+    client = get_s3_client(meta)
+    path_list = src.lstrip('/').split('/', 1)
+    if len(path_list) == 1:
+        logger.error("failed", task="irods_upload_S3file", path=src)
```

Do we need to bail here? What does the error mean if it's allowed to continue?

---

Yes, I would think we need to bail. If the length of the path list is less than 1, it means that the source_physical_fullpath either does not contain the filename within the S3 bucket, or does not contain the bucket name. If we continue, it would lead to an error when trying to get the object from the S3 bucket via the Minio client. 

What does it look like to bail from the upload file? I'm not sure what needs to happen.

---

I think we need to raise an Exception. I think a `ValueError` would be appropriate here based on your description.

On that note, we should definitely leave a more detailed message here than "failed" as well. Perhaps something to the effect of your explanation would suffice.